File: 2019/TypeScript/01.ts
import * as fs from "fs";
import * as path from "path";

function readInput(): string {
  const inputPath = path.join(__dirname, "../input/01.txt");
  return fs.readFileSync(inputPath, "utf8").trim();
}

function part1(input: string): number {
  let fuel = 0;

  for (const mass of input.split("\r\n")) {
    fuel += Math.floor(Number(mass) / 3) - 2;
  }

  return fuel;
}

function part2(input: string): number {
  let fuel = 0;

  for (const mass of input.split("\r\n")) {
    let tmpFuel = Math.floor(Number(mass) / 3) - 2;

    while (tmpFuel > 0) {
      fuel += tmpFuel;
      tmpFuel = Math.floor(tmpFuel / 3) - 2;
    }
  }

  return fuel;
}

const input = readInput();

console.log("Part 1:", part1(input));
console.log("Part 2:", part2(input));

File: 2019/Rust/01.rs
mod day_01 {
    use std::fs;
    use std::path::{Path, PathBuf};

    fn read_input() -> String {
        let input_path: PathBuf = Path::new(file!()).parent().unwrap().join("../input/01.txt");
        fs::read_to_string(input_path)
            .expect("Failed to read input file")
            .trim()
            .to_string()
    }

    fn part1(input: &str) -> i32 {
        input
            .lines()
            .map(|mass| {
                let fuel = mass.parse::<i32>().unwrap();
                (fuel / 3) - 2
            })
            .sum()
    }

    fn part2(input: &str) -> i32 {
        let mut fuel = 0;

        for mass in input.split("\r\n") {
            let mut tmp_fuel = (mass.parse::<i32>().unwrap() / 3) - 2;

            while tmp_fuel > 0 {
                fuel += tmp_fuel;
                tmp_fuel = (tmp_fuel / 3) - 2;
            }
        }

        fuel
    }

    pub fn run() {
        let input: String = read_input();

        println!("Part 1: {}", part1(&input));
        println!("Part 2: {}", part2(&input));
    }
}

pub use day_01::run;
